### PR TITLE
Feature/tests

### PR DESCRIPTION
**Pull request by mohamednaser**
we finished the test cases for the enumerable project on three files 

- [x] enumerable_conditionals file

- [x] enumerable_loops file

- [x] enumerable_misc file

and we are waiting for your review to merge with the develop branch 

thanks in advance

### CODE REVIEW

**Comment by santiago-rodrig** [comment]
# Dear reviewer

The changes required are made. I think that this project is already done, but that is up to you. A list of the major changes in this feature branch follows.

 - [x] Tests for **iteration methods**, this is for `my_each` and `my_each_with_index`. The tests for both of these can be found in the `spec/loops_spec.rb` file.
 - [x] Tests for **conditional methods**, this is for `my_all?`, `my_none?`, and `my_any?`. The tests for these can be found in the files `spec/conditionals/no_block_spec.rb`, `spec/conditionals/block_no_args.rb`, and `spec/conditionals/block_and_args.rb`.
 - [x] Tests for **miscelaneous** methods, this is for `my_map`, `my_select`, `my_count`, and `my_inject`. The tests for these can be found in the `spec/misc_spec.rb` file.

Here I'll leave the instructions on how to run the tests **just in case**.

[asciicast](https://asciinema.org/a/280095)

File: spec/conditionals/block_and_args_spec.rb [hunk 1]
```diff
@@ -0,0 +1,1220 @@
...
+              it 'should return true' do
+                # regex, empty array
+                expect([].my_all?(/./) { false }).to be_truthy
+                expect([].my_all?(/./) { |foo| foo }).to be_truthy
+                expect([].my_all?(/./) { |foo, bar| bar }).to be_truthy
+                expect([].my_all?(/./) { |foo, bar, baz| baz }).to be_truthy
+                expect([].my_all?(/\d\w\s[a-z0-9]/) { false }).to be_truthy
+                expect([].my_all?(/\d\w\s[a-z0-9]/) { |foo| foo }).to be_truthy
+                expect([].my_all?(/\d\w\s[a-z0-9]/) { |foo, bar| bar }).to be_truthy
+                expect([].my_all?(/\d\w\s[a-z0-9]/) { |foo, bar, baz| baz }).to be_truthy
```
Comment (bolah2009) [inline spec/conditionals/block_and_args_spec.rb]: Don't you think this is too much for only one `it` block? :thinking: Most of what we testing here are redundant.

#### A reminder from [Pathwright Project specific](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59566098/)

> Use one `it` block for each expected behaviour. **Do not test multiple aspects in one it block**.

File: spec/misc_spec.rb [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+          expect((0..-1).my_inject { |sum, n| sum + n }).to eq(nil)
+        end
+
+        it 'return value of applying binary operation spesified by block ' do
+          expect(range.my_inject { |sum, n| sum + n }).to eq(55)
+        end
+      end
+    end
+
+    describe '#my_select' do
```
Comment (bolah2009) [inline spec/misc_spec.rb]: Could we add test for when the enumerable is a `Hash`? The test still passes while commenting out this (👇🏾) part of the code.

```ruby  
def my_select
    return to_enum :my_select unless block_given?

    # if is_a? Hash
    #   selection = {}
    #   my_each { |k, v| selection[k] = v if yield k, v }
    #   return selection
    # end
    selection = []
    my_each { |v| selection << v if yield v }
    selection
  end

File: spec/misc_spec.rb [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+          end
+
+          it 'return zero when there is no matches ' do
+            expect(array.my_count(99)).to eq(0)
+          end
+        end
+        context 'range' do
+          it 'return number of matches for the argument ' do
+            expect(range.my_count(2)).to eq(1)
+          end
```
Comment (bolah2009) [inline spec/misc_spec.rb]: Could we add a test for when a block is given? The test still passes while commenting out this (👇🏾) part of the code.

```ruby
  def my_count(*args)
    counter = 0
    unless args.empty?
      arg = args[0]
      to_a.my_each { |v| counter += 1 if v == arg }
      return counter
    end
    # if block_given?
    # case block.parameters.size
    # when 1
    #   to_a.my_each { |v| counter += 1 if block.call(v) }
    # when 2
    #   my_each { |k, v| counter += 1 if block.call(k, v) } if is_a? Hash
    # end
    # else
    counter = size
    # end
    counter
  end
```

File: spec/misc_spec.rb [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+            expect { array.my_map(2) { |ele| p ele } }.to raise_error(ArgumentError)
+          end
+        end
+
+        context 'Range' do
+          it 'in pass an argument raise an error ' do
+            expect { (0..-1).my_map(2) { |ele| p ele } }.to raise_error(ArgumentError)
+          end
+        end
+      end
```
Comment (bolah2009) [inline spec/misc_spec.rb]: Check this commented out part of `my_map` too.

```ruby
  def my_map(&block)
    return to_enum :my_map unless block_given?

    mapped = []
    case block.parameters.size
    when 1
      to_a.my_each { |v| mapped << block.call(v) }
    when 2
      # my_each { |k, v| mapped << block.call(k, v) } if is_a? Hash
    end
    mapped
  end
```

File: .rubocop.yml [hunk 5]
```diff
@@ -18,7 +18,8 @@ Metrics/AbcSize:
 Metrics/ClassLength:
   Max: 150
 Metrics/BlockLength:
-  ExcludedMethods: ['describe']
+  Max: 200
+  ExcludedMethods: ['describe', 'context', 'it']
```
[the remark this answers is not in the document]
Comment (santiago-rodrig) [reply .rubocop.yml]: Thanks for the hint!

File: README.md [hunk 6]
```diff
@@ -1,10 +1,14 @@
 ﻿# Enumerable methods
 
-This proyect has as a goal the recreation of some of the methods provided by the [Enumerable](https://ruby-doc.org/core-2.6.3/Enumerable.html) module. Project instructions are in [The Odin Project](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#project-2-enumerable-methods).
+This proyect has as a goal the recreation of some of the methods provided by the
```
[the remark this answers is not in the document]
Comment (santiago-rodrig) [reply README.md]: Oh, now is fixed, thanks.

File: spec/conditionals/block_and_args_spec.rb (hunk 1 above)
Comment (santiago-rodrig) [reply spec/conditionals/block_and_args_spec.rb]: I'll simplify everything.

File: spec/my_inject_spec.rb [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+    context 'a range' do
+      let(:collection) { range }
+
+      it 'returns the block returned value' do
+        should eq('banzai')
+      end
+    end
+  end
+
+  context 'with one argument and a block of 1 variable' do
```
[the remark this answers is not in the document]
Comment (santiago-rodrig) [reply spec/my_inject_spec.rb]: solved

**Pull request merged**